manifests/init.pp

```diff
@@ -0,0 +1,130 @@
+# @summary Manage a MySQL/MariaDB server configuration.
+#
+# @param create_mysql_home
+# Should the mysql_home be created?
+#
+# @param dbs
+# Raw data to define mysql::db resources.
+#
+# @param mysql_gid
+# The GID for the mysql user.
+#
+# @param mysql_groupname
+# The name of the mysql group.
+#
+# @param mysql_home
+# The home dir of the mysql user.
+#
+# @param mysql_logdir
+# Location of MySQL log files.
+#
+# @param mysql_parentdir
+# Location of directory that contains the mysql_home.
+# NOTE: Might be useful for this to be an array so multiple nested
+# dirs can be created as dependencies.
+#
+# @param mysql_rundir
+# Location of run directory (generally a sub-directory of /var/run/).
+#
+# @param mysql_uid
+# UID of the mysql user.
+#
+# @param mysql_username
+# Username of the mysql user.
+#
+# @param other_dependencies
+# Other depencies to handle prior to other MySQL setup, specified as resources, e.g.:
+#   - "Mount['/var/lib/mysql']"
+#
+# @param yumrepo
+# Raw params for a yumrepo resource from which to install MySQL/MariaDB.
+#
+# @example
+# include profile_mysql_server
+#
+class profile_mysql_server (
+
+  Boolean           $create_mysql_home,
+  Hash              $dbs,
+  Integer           $mysql_gid,
+  String            $mysql_groupname,
+  String            $mysql_home,
+  String            $mysql_logdir,
+  Optional[String]  $mysql_parentdir,
+  String            $mysql_rundir,
+  Integer           $mysql_uid,
+  String            $mysql_username,
+  Array             $other_dependencies,
+  Hash              $yumrepo,
+
+) {
+
+  include ::mysql::server
+
+  $dir_defaults = {
+    before => Class['::mysql::server::install'],
+    ensure => directory,
+    group  => $mysql_gid,
+    owner  => $mysql_uid,
+  }
+
+  if $mysql_logdir {
+    file { $mysql_logdir:
+      * => $dir_defaults,
+    }
+  }
+  if $mysql_rundir {
+    file { $mysql_rundir:
+      * => $dir_defaults,
+    }
+  }
+  if $mysql_parentdir {
+    file { $mysql_parentdir:
+      * => $dir_defaults,
+    }
+  }
+
+  if $create_mysql_home {
+    file { $mysql_home:
+      * => $dir_defaults,
+    }
+  }
+
+  $other_dependencies.each | $dep | {
+    $dep -> Class['::mysql::server::install']
+  }
+
+  group { $mysql_groupname:
+    ensure => 'present',
+    before => Class['::mysql::server::install'],
+    gid    => $mysql_gid,
+  }
+
+  user { $mysql_username:
+    ensure         => 'present',
+    before         => Class['::mysql::server::install'],
+    uid            => $mysql_uid,
+    gid            => $mysql_gid,
+    forcelocal     => true,
+    home           => $mysql_home,
+    managehome     => $create_mysql_home,
+    password       => '!!',
+    purge_ssh_keys => true,
+    shell          => '/sbin/nologin',
+    comment        => 'MySQL server',
+  }
+
+  if ! empty($yumrepo) {
+    yumrepo { 'mysql_yumrepo':
```

Comment:
Should this instead use ensure_resources (‘yumrepo’, $yumrepo) - just in case the puppet/mysql or another module try to ensure the same thing?

Reply:
I added a $manage_yumrepo param that defaults to false. If $yumrepo is defined in Hiera (rather than just falling back to the data-in-module value of {}) and is managed by Puppet (either this module or another), then Puppet will ensure it happens first. Additionally if $create_yumrepo is true then this module will ensure it.